Add the ability to have user defined features via a PlugInFeature class

PinkShnack opened the pull request:
This is described in Issue #105. The up to date layout of this PR is copied below. Fixes #105 


**Possible layout of the PlugInFeature and user script workflow:**

    (dclab) RTDCBase
                  |
                  | checks for ...
                  |
    (dclab) AncillaryFeature class
                  |
                  | which is the parent class of ...
                  |
    (dclab) PlugInFeature class
                  |
                  | whose methods can be overridden by those defined by the user in the ...
                  |
    (ext. Script) ExamplePlugin(PlugInFeature) class
        What the user interacts with.
        The user defines what is needed to calculate the new feature, and how to calculate it (like with Ancillary)


### Checklist
- [x] `PlugInFeature` class inherits from `AncillaryFeature` class
- [x] Functions to import user script and initialize `PlugInFeature`
- [x] Add tests for `PlugInFeature` and tests for imported `PlugInFeature` 
- [x] Mirror suitable tests from `test_temporary_features.py`
- [x] Add documentation for website
- [x] Correct docstrings for all new and changed functions/tests/methods
  - [x] Add docstrings for `plugin_feature.py` and `dclab.dfn`
  - [x] Add docstrings for plugin tests
- [x] Create tests for minimum `PlugInFeature` `info` allowed (just name and method)

### Bugs and ideas
- Create tests for new `dclab.dfn` functions. See #121 
- When running tests for `AncillaryFeature.data` attribute, noticed that it isn't updated #122 
- Bug: fix contour loading issue, and remove `dclab.dfn.check_feature_shape` if statement for `"contour"`  See #117

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,27 @@
+    def __init__(self, *args, **kwargs):
+        super(PlugInFeature, self).__init__(*args, **kwargs)
+
+        self.feature_name = self.get_feature_name
```
This method should be run so to return the `str` i.e., `self.feature_name = self.get_feature_name()`

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_example.py:
```diff
@@ -0,0 +1,41 @@
+
+
+    def get_feature_name(self):
+        ''' Compute the feature you want
```
Change this, and format docstrings correctly.

PinkShnack commented:
Hey @paulmueller, @maxschloegel and I had a discussion about this PR/Issue. 
Both you feel free to pull the branch and make changes!

Paul mentioned that we should tackle #104 first. Perhaps, seeing as I still don't understand what needs to be done there, we could all have a chat about it and the above code.

paulmueller reviewed (commented):
I made a more elaborate comment in https://github.com/ZELLMECHANIK-DRESDEN/dclab/issues/105#issuecomment-805306060 to illustrate why I brought #104 into play.

PinkShnack commented on dclab/__init__.py:
```diff
@@ -15,6 +15,8 @@
 from .rtdc_dataset import new_dataset
 from .rtdc_dataset.feat_temp import register_temporary_feature, \
```
I guess there is no difference between how to do line continuations for imports, but PEP328 says to use parentheses (and I like it :) ) https://www.python.org/dev/peps/pep-0328/
If you agree then I can go through the code and do this everywhere. Or we can get @maxschloegel to write that python formatting package we mentioned ;)

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,65 @@
+from ..ancillaries import AncillaryFeature
+
+
+_registered_plugin_instances = []
```
See the `remove_all_plugin_features` function on line 49.

paulmueller replied:
I think we don't necessarily need that. All plugin features are instances of `PluginFeature` and are listed in `AncillaryFeature.features`.

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,65 @@
+        ref_info = {
+            "feature_name": info["feature names"][ii],
+            "method": info["method"],
+            "req_config": info["config required"][ii],
```
Both `"req_config"` and `"req_features"` need to be list of lists for propper hashing, see `examples/plugin_example_features.py`. It isn't the cleanest look.

PinkShnack commented on examples/plugin_example_features.py:
```diff
@@ -0,0 +1,20 @@
+    "feature names": ["circ_per_area", "bright_per_area"],
+    "feature labels": ["", ""],
+    "features required": [["circ", "area_um"], ["bright_avg", "area_um"]],
+    "config required": [[], []],
```
as mentioned above, not the cleanest implementation.

paulmueller replied:
See my comments above.

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,65 @@
+"""
+.. versionadded:: 0.34.0
```
Hopefully anyway

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,65 @@
+        example plugins.
+        """
+        super().__init__(**kwargs)
+        self.plugin_registered = True
```
I'm sure we could have some nice attributes here, ideas?

paulmueller replied:
- `author`
- `version`
- `description`
- `long description`

EDIT: or just leave everything in `plugin_info`. Fine with me.

PinkShnack commented on tests/test_plugin_features.py:
```diff
@@ -0,0 +1,100 @@
+from helper_methods import retrieve_data
+
+
+@pytest.fixture
```
Currently a placeholder for a fixture similar to your temporary feature fixture.

paulmueller replied:
Yes, just call `remove_all_plugin_features()` at the bottom and set `autouse=True` so every test automatically gets it.

PinkShnack commented on tests/test_plugin_features.py:
```diff
@@ -0,0 +1,100 @@
+    PlugInFeature.features.remove(pf)
+    PlugInFeature.feature_names.remove(pf.feature_name)
+    ds3 = dclab.new_dataset(retrieve_data("rtdc_data_hdf5_rtfdc.zip"))
+    assert "circ_per_area" not in ds3
```
I got paranoid after a while. There may be a nice way to do this with a `with` context manager.

paulmueller replied:
Cleanup can be done by the fixture above.

PinkShnack commented:
Regarding the changes uploaded today (27.04.21)

Here is how one might add their nice new `PlugInFeature` for use with calculations @paulmueller.
It works! But whether it is correct and clean is another question!


```python

import dclab

ds = dclab.new_dataset("path/to/rtdc/file")

# the below path could be standard on their computer.
plugin_list = dclab.create_new_plugin_feature(plugin_path="path/to/dclab/examples/plugin_example_features.py")

circ_per_area = ds["circ_per_area"]
bright_per_area = ds["bright_per_area"]

# do some filtering etc.
```

paulmueller commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -1,27 +1,65 @@
+
+def remove_plugin_feature(plugin_instance):
+    """Convenience function for removing a plugin instance"""
+    PlugInFeature.features.remove(plugin_instance)
```
Please add a check that `plugin_instance` is an instance of `PluginFeature`. And in a test, make sure that removing an `AncillaryFeature` does not work.

paulmueller commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -1,27 +1,65 @@
-    def get_feature_name(self):
-        pass
+def remove_all_plugin_features():
+    # I guess because we are built on AncillaryFeature, we don't need something
```
Instead, here you could loop over all `AncillaryFeature.features` and remove all instances of `PluginFeature`.

PinkShnack replied:
Thought so! Great

paulmueller commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -1,27 +1,65 @@
+        example plugins.
+        """
+        super().__init__(**kwargs)
+        self.plugin_registered = True
```
What is `self.plugin_registered` there for? I think we can remove it since there is no scenario where no exception is raised and this is `False`. Also, it will never be `False` anyway. Deactivating a plugin should be done by removing it from the `AncillaryFeature.features` list.

paulmueller commented on examples/plugin_example_features.py:
```diff
@@ -0,0 +1,20 @@
+    "method": compute_some_new_features,
+    "description": "This plugin will compute some possibly useless features",
+    "feature names": ["circ_per_area", "bright_per_area"],
+    "feature labels": ["", ""],
```
`feature labels` should never be empty (Please add a check for that during instantiation). It can be omitted though (in which case it should default to `feature names`).

PinkShnack replied:
`feature_labels` are added to the `dfn` file, right? Like for temporary features?

paulmueller replied:
No, we have to do it in a similar way as for the temporary features. Ancillary features don't need that, because they are only defined for features that are in the `dfn` file.

paulmueller replied:
I think it would make sense to have a shared code base for registering and deregistering plugin and temporary features in the `dfn` file.

paulmueller commented on examples/plugin_example_features.py:
```diff
@@ -0,0 +1,20 @@
+    "feature labels": ["", ""],
+    "features required": [["circ", "area_um"], ["bright_avg", "area_um"]],
+    "config required": [[], []],
+    "method check required": "",
```
This should be a callable. If it is omitted it should default to None (not sure how it is handled in `AncillaryFeature` currently.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -0,0 +1,100 @@
+    circ_per_area = ds["circ_per_area"]
+    assert np.allclose(circ_per_area, ds["circ"] / ds["area_um"])
+
+    assert "circ_per_area" in ds
```
This assertion should come before line 48, otherwise it is redundant.

PinkShnack replied:
Nice thanks

paulmueller reviewed (changes requested):
I think a test that actually imports a plugin-feature from a .py file should also be added.

paulmueller commented:
I think I would rename the functions and change the behavior a little. The thing is that each plugin should be an individual file. To take your example:

```python
import dclab

# load a single plugin feature
dclab.load_plugin_feature("/path/to/plugin_feature.py")

# I am not sure whether we should support something like this:
dclab.load_plugin_features_from_folder("/path/to/folder_containing_plugin_features/", recursive=True)
# The problem with this is that this potentially leads to many errors if one plugin does not work and the user is overwhelmed.

# load some data (even though it will also work the other way around, I would recommend users to
# load the plugin features at the beginning of their scripts.
ds = dclab.new_dataset("path/to/rtdc/file")
# access the features
circ_per_area = ds["circ_per_area"]
bright_per_area = ds["bright_per_area"]

# do some filtering etc.
```
I think dclab should not have any default directories for plugins. This should be handled by higher-level software like Shape-Out that can then also depend on e.g. `appdirs` or similar.

paulmueller commented:
Another important point would be a section on plugin-features in the docs.

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -1,27 +1,65 @@
+    # find script, return info dict
+    path = pathlib.Path(plugin_path)
+    # insert the plugin directory to sys.path so we can import it
+    sys.path.insert(-1, str(path.parent))
```
> # I am not sure whether we should support something like this:
> dclab.load_plugin_features_from_folder("/path/to/folder_containing_plugin_features/", recursive=True)
> # The problem with this is that this potentially leads to many errors if one plugin does not work and the user is overwhelmed.

This would be really nice. I feel like it wouldn't be difficult for `glob` to handle. I will see how many errors it throws up...


> I think dclab should not have any default directories for plugins. This should be handled by higher-level software like Shape-Out that can then also depend on e.g. `appdirs` or similar.

I agree, I was more hinting that we could tell users to have one place for all their plugins. Autoload with ShapeOut2 would be awesome.

I forgot to mention that I will be adding many more tests. Do you think loading the `examples/plugin.py` file would be okay to do for testing, or will I include a plugin in the test folder?

PinkShnack replied:
It seems to not work without modifying the `sys.path`. Also, as far as I understand `finally` is always excecuted, so we should not `pop()` in the `finally` clause.

PinkShnack commented:
> dclab.load_plugin_features_from_folder("/path/to/folder_containing_plugin_features/", recursive=True)


This would be really nice. I feel like it wouldn't be difficult for `glob` to handle. I will see how many errors it throws up...


> I think dclab should not have any default directories for plugins. This should be handled by higher-level software like Shape-Out that can then also depend on e.g. `appdirs` or similar.

I agree, I was more hinting that we could tell users to have one place for all their plugins. Autoload with ShapeOut2 would be awesome.

I forgot to mention that I will be adding many more tests. Do you think loading the `examples/plugin.py` file would be okay to do for testing, or will I include a plugin in the test folder?

paulmueller commented:
> > dclab.load_plugin_features_from_folder("/path/to/folder_containing_plugin_features/", recursive=True)
> 
> This would be really nice. I feel like it wouldn't be difficult for `glob` to handle. I will see how many errors it throws up...

I think this is why we should let the user do this (instead of having it as a function in dclab):
```python
for plugin_path in pathlib.Path("my_plugin_directory").rglob("*.py"):
    dclab.load_plugin_feature(plugin_path)
```

> I forgot to mention that I will be adding many more tests. Do you think loading the examples/plugin.py file would be okay to do for testing, or will I include a plugin in the test folder?

It would be cleaner to have a separate plugin in the `tests/data` folder.

paulmueller commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,66 @@
+    """Convenience function for removing a plugin instance"""
+    if isinstance(plugin_instance, PlugInFeature):
+        PlugInFeature.features.remove(plugin_instance)
+        PlugInFeature.feature_names.remove(plugin_instance.feature_name)
```
Please add en `else` clause that raises an error. This function should not silently fail.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -0,0 +1,136 @@
+    remove_all_plugin_features()
+
+
+def get_plugin_file(plugin_name="plugin_test_example.py"):
```
To be consistent with the other tests, I would remove this function and use `data_dir=pathlib.Path(...)` at the top of the file and in the test function I would then use`plugin_path = data_dir / "plugin_test_example.py"`.

PinkShnack replied:
Cool, will do. I thought it'd be a bit cleaner this way, similar to what you had done in `test_isoelastics.py`:
```python
def get_isofile(name="example_isoelastics.txt"):
    thisdir = pathlib.Path(__file__).parent
    return thisdir / "data" / name
```

If we want to be consistent, shall I change these kinds of functions everywhere to the `data_dir=...` code?

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -0,0 +1,136 @@
+    plugin_list = dclab.load_plugin_feature(plugin_path)
+    assert isinstance(plugin_list[0], PlugInFeature)
+    assert isinstance(plugin_list[1], PlugInFeature)
+
```
another critical assertion is `assert dclab.dfn.feature_exists("circ_per_area")`.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -0,0 +1,136 @@
+    assert "circ_per_area" in ds
+    assert "circ_times_area" in ds
+
+    remove_all_plugin_features()
```
```python
assert  dclab.dfn.feature_exists("circ_per_area")
remove_all_plugin_features()
assert not dclab.dfn.feature_exists("circ_per_area")
```

paulmueller commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,66 @@
+        plugin = importlib.import_module(path.stem)
+        # undo our path insertion
+        sys.path.pop(0)
+
```
I think it might make sense to add `plugin_path` to `plugin.info`. Could be useful for debugging.

PinkShnack replied:
Btw, I took this to mean let's add `plugin_path` as a `PlugInFeature` class attribute. I think this is useful because you can be sure where it is being loaded from.

Did you mean to add it to the `info` dict when writing the plugin script? Example:

```python
info = {
    ...
    "priority": 1,
    "plugin_path": 'path/to/file',
    "version": "0.1.0",
}
```
I can imagine this being defunc quite quickly as the user may move the plugin script around/onto a different computer etc.

paulmueller replied:
No, "plugin_path" should not be specified by the user - dclab should define it. Maybe it makes sense to have it as an attribute as opposed to in the `plugin_info` dictionary (because this might be confusing).

PinkShnack replied:
Great I was on the right track. And yes it is its own attribute.

paulmueller reviewed (changes requested):
Please also have a look at https://github.com/ZELLMECHANIK-DRESDEN/dclab/blob/master/tests/test_temporary_features.py and see which tests you can adapt for plugin features. Please add a test with `dclab.dfn.get_feature_label("circ_per_area")`.

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -0,0 +1,66 @@
+        plugin = importlib.import_module(path.stem)
+    except ModuleNotFoundError:
+        sys.path.insert(-1, str(path.parent))
+        plugin = importlib.import_module(path.stem)
```
I think I have done this correctly now. Actually never really needed try except very much!

PinkShnack commented on dclab/definitions.py:
```diff
@@ -364,3 +365,52 @@ def get_feature_label(name, rtdc_ds=None):
+
+
+def update_dfn_with_feature_info(feature, label, is_scalar):
+    """Used by temporary features and plugin features to update the feature
```
This and the two following functions can go anywhere, I thought here in the definitions file would be best. Should we make them protected i.e., `def _update_dfn_with_feature_info(...):` ?

PinkShnack commented:
I have discovered what may be a bug in `contour.shape`. This may be related to #117. 
I have pushed a quick/probably not permanent fix in this commit: ["add shared code base for checking shape of features. Added if statement for contour bug: d9f6352"](https://github.com/ZELLMECHANIK-DRESDEN/dclab/pull/113/commits/d9f63522b4ad30bf190df513ae52af361466c650) with this code:

```python
def check_feature_shape(feature, data):
    data = np.array(data)
    # bug: reloaded contour is of the form: array([array([[1,2],[...]...])])
    # which means len(data.shape) == 1.
    if feature == "contour" and len(data.shape) == 1:
        pass
    elif len(data.shape) == 1 and not scalar_feature_exists(feature):
        ...
```

Here is a standard traceback:

```bash
    def test_split_traces():
        path_in = retrieve_data("rtdc_data_hdf5_rtfdc.zip")
>       paths = cli.split(path_in=path_in, split_events=3, ret_out_paths=True)

tests\test_cli.py:243:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dclab\cli.py:517: in split
    skip_empty_image_events(ds, initial=skip_initial_empty_image,
dclab\cli.py:423: in skip_empty_image_events
    or ("contour" in ds and np.all(ds["contour"][0] == 0))
dclab\rtdc_dataset\core.py:110: in __getitem__
    data_dict = ancol[key].compute(self)
dclab\rtdc_dataset\ancillaries\ancillary_feature.py:208: in compute
    dfn.check_feature_shape(self.feature_name, data_dict[key])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

feature = 'contour'
data = array([array([[186,  36],
       [186,  35],
       [186,  34],
       [186,  33],
       [185,  32],
       [185,  31...,  45],
       [107,  44],
       [108,  43],
       [109,  42],
       [109,  41],
       [110,  40]])], dtype=object)

    def check_feature_shape(feature, data):
        data = np.array(data)
        # bug: reloaded contour is of the form: array([array([[1,2],[...]...])])
        # which means len(data.shape) == 1.
        # if feature == "contour" and len(data.shape) == 1:
        #     pass
        if len(data.shape) == 1 and not scalar_feature_exists(feature):
>           raise ValueError("Feature '{}' is not a scalar feature, but a "
E           ValueError: Feature 'contour' is not a scalar feature, but a 1D array was given for `data`!

```

paulmueller commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -6,12 +6,17 @@
+    pass
+
+
 def load_plugin_feature(plugin_path):
```
You can use e.g. 
```python
if "scalar feature" not in info:
    info["scalar feature"] = [True] * len(info["feature names"])
```

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -21,13 +28,143 @@ def cleanup_plugin_features():
+    """Test for RTDCHierarchy"""
+    plugin_path = ''
+    other_info = {}
+    feature_label = ""
```
Feature labels should never be empty strings (this should not work). They may be `None`, but then `dfn` will take care of it. Initially, this should be handled by `load_plugin_feature`.

PinkShnack replied:
As discussed earlier, I have a check for this. IF the user for whatever reason leaves labels as empty strings, they will be set to None, and then the dfn file will handle it as usual.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -41,55 +178,40 @@ def test_create_plugin():
+        feature_info = example_plugin_feature_info()
+        PlugInFeature(feature_label, is_scalar, plugin_path,
+                      other_info, **feature_info)
+        assert "circ_per_area" in ds
```
also: `assert "circ_per_area" in ds.features_innate`

PinkShnack replied:
Bit confused here. do you mean `assert "circ_per_area" not in ds.features_innate`?

paulmueller replied:
I probably meant *before* creation of the plugin.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -101,32 +223,165 @@ def test_single_plugin_feature():
-            "priority": 1,
-        }
-        plugin_list.append(PlugInFeature(other_info, **feature_info))
+    feature_labels = ["Circularity per Area", "Circularity times Area"]
```
Also check that `dfn` returns the correct feature labels.

paulmueller commented on tests/test_plugin_features.py:
```diff
@@ -101,32 +223,165 @@ def test_single_plugin_feature():
+
+
+def test_pf_wrong_length_1():
+    """temporary feature should have same length"""
```
Docstrings still have "temporary feature" in it.

paulmueller reviewed (changes requested):
Tests are very thorough - thank you!
I made a few suggestions that would simplify a few things and support writing shorter plugins by leaving out certain `info` keywords.

PinkShnack commented on dclab/definitions.py:
```diff
@@ -364,3 +365,52 @@ def get_feature_label(name, rtdc_ds=None):
+
+def check_feature_shape(feature, data):
+    data = np.array(data)
+    # bug: reloaded contour is of the form: array([array([[1,2],[...]...])])
```
Details noted in #117

PinkShnack commented on dclab/rtdc_dataset/plugins/plugin_feature.py:
```diff
@@ -45,8 +55,15 @@ def find_plugin_feature_script(plugin_path):
 def remove_plugin_feature(plugin_instance):
     """Convenience function for removing a plugin instance"""
     if isinstance(plugin_instance, PlugInFeature):
+        if plugin_instance.feature_name in PlugInFeature.feature_names:
```
Writing here for clarity: 
This will not always be `True` because the `PlugInFeature` setup happens in this order:
1. `PlugInFeature` class is initialised
2. The `feature_label`, `feature_name` etc are checked. If they fail, then you still have a `PlugInFeature` instance.

Therefore, the `PlugInFeature` instance can exist without a corresponding `feature_label` and `feature_name`.

If you try to delete these attributes without this check, it will most of the time just throw an error. However, if the user sets their `feature_name="deform"` and then `remove_plugin_feature` is run, `"deform"` will be removed from the `dclab.dfn` file!

paulmueller commented:
Could you please rebase this PR on master and make sure the changelog is correct? There is no 0.33.4 release, so that should be removed from the changelog.

paulmueller commented on examples/plugin_load_script.py:
```diff
@@ -0,0 +1,36 @@
+# load a single plugin feature
+dclab.load_plugin_feature("/path/to/plugin_example_features.py")
+
+# loading multiple plugins
```
It is fine that his is just in the documentation.

paulmueller commented on examples/plugin_load_script.py:
```diff
@@ -0,0 +1,36 @@
+    dclab.load_plugin_feature(plugin_path)
+
+# load some data
+ds = dclab.new_dataset("path/to/rtdc/file")
```
You can use DCOR data here.

paulmueller commented on examples/plugin_load_script.py:
```diff
@@ -0,0 +1,36 @@
+ds.apply_filter()
+print("Removed {} out of {} events!".format(np.sum(~ds.filter.all), len(ds)))
+
+# save the plugin features in a file
```
Instead of exporting to a file (which is already in the documentation), how about creating a nice plot for this example?

PinkShnack replied:
Check out my lame plot... https://github.com/PinkShnack/dclab/blob/plugin_feature_1/examples/plugin_plot.jpeg
Perhaps there is something nicer that could be shown?

paulmueller replied:
It does not really matter as long as it demonstrates plugin features. I would change/remove the title. Please note that for reproducibility, the plot should be generated with this script: https://github.com/ZELLMECHANIK-DRESDEN/dclab/blob/master/examples/generate_example_images.py
Have a look at the other examples to see how I implemented it.

paulmueller commented on examples/plugin_load_script.py:
```diff
@@ -0,0 +1,36 @@
+
```
This example does not have a docstring.

paulmueller commented on tests/data/plugin_test_example.py:
```diff
@@ -0,0 +1,22 @@
+
```
There is no docstring for this example

paulmueller reviewed (changes requested):
I made a few suggestions for the doc section. You should always assume that the user is completely clueless. Always start with a general description of what this is and what it could be used for. If the user is still interested he will read on and you can explain it in more detail.
